api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/summary/FormatUtils.kt

```diff
@@ -0,0 +1,261 @@
+    }
+
+    override fun getWidth(): Int {
+        return sectionWidth
+    }
+}
+
+/**
+ * Section with columns of the model summary table.
+ */
+public class SectionWithColumns(
+    private val rows: List<TableRow>,
+    private val columnNames: List<String>,
+    private val columnSeparator: String,
+    private val lineSeparatorSymbol: Char,
+    private val thickLineSeparatorSymbol: Char
+) : Section {
+    private var columnsWidths = calcColumnWidths(rows, columnNames)
+
+    override fun format(width: Int): List<String> {
+        require(width >= getWidth()) { "Section width must be greater than or equal to ${getWidth()}" }
+        updateColumnWidths(width)
+
+        val lineSeparator = lineSeparatorSymbol.toString().repeat(width)
+        val thickLineSeparator = thickLineSeparatorSymbol.toString().repeat(width)
+
+        val result = mutableListOf<String>()
+
+        result.add(formatLine(columnSeparator, columnsWidths, columnNames))
+
+        result.add(thickLineSeparator)
+
+        rows.forEach { row ->
+            result.addAll(row.format(columnSeparator, columnsWidths))
+            result.add(lineSeparator)
+        }
+
+        return result
+    }
+
+    override fun alignWith(anotherSection: Section) {
+        if (anotherSection == this) return
+        if (anotherSection is SimpleSection) return
+
+        anotherSection as SectionWithColumns
+
+        when {
+            columnsWidths.size >= anotherSection.columnsWidths.size -> {
+                anotherSection.columnsWidths.forEachIndexed { index, columnWidth ->
+                    columnsWidths[index] = max(columnWidth, columnsWidths[index])
+                }
+            }
+
+            else -> {
+                columnsWidths.forEachIndexed { index, columnWidth ->
+                    columnsWidths[index] = max(columnWidth, anotherSection.columnsWidths[index])
+                }
+
+                val extraColumnsWidths = anotherSection.columnsWidths
+                    .subList(columnsWidths.size, anotherSection.columnsWidths.size)
+
+                columnsWidths[columnsWidths.lastIndex] +=
+                    extraColumnsWidths.sum() + extraColumnsWidths.size * columnSeparator.length
+            }
+        }
+    }
+
+    override fun alignWith(anotherSections: List<Section>) {
+        for (section in anotherSections) {
+            alignWith(section)
+        }
+    }
+
+    override fun getWidth(): Int {
+        return columnsWidths.sum() + (columnsWidths.size - 1).coerceAtLeast(0) * columnSeparator.length
+    }
+
+    private fun updateColumnWidths(newWidth: Int) {
+        if (newWidth == getWidth()) return
+
+        columnsWidths = columnsWidths.map { it + (newWidth - getWidth()) / columnsWidths.size }.toMutableList()
+    }
+}
+
+
+/**
+ * The row of the summary table may consist of multiple lines.
+ * This class is used to simplify the handling of formatting of the rows.
+ */
+public class TableRow(lines: List<List<String>>) {
```

Comment:
I would expect that a row consists of cells, and each cell can have a multiple lines of text. This would simplify creating rows in the model code.